### ISSUE TITLE
logind: reboot on KEY_RESTART

### ISSUE THREAD

**Issue by robimarko** [systemd/systemd#16400]
**systemd version the issue has been seen with**
> 244.3

**Used distribution**
 > Gentoo based

**Expected behaviour you didn't see**
 > udevadm monitor sees no events for gpio based buttons.
I have reset button that is a simple GPIO driven by the gpio-keys driver which is enabled along the HID and input subsystems,
but udev sees no events when using the button so a rule to use it to reboot cant be set.
Kernel normally sees the change and creates the KEY_RESTART code event:
```
[   45.952802] evbug: Event. Dev: input0, Type: 1, Code: 408, Value: 1
[   45.952914] evbug: Event. Dev: input0, Type: 0, Code: 0, Value: 0
[   46.922746] evbug: Event. Dev: input0, Type: 1, Code: 408, Value: 0
[   46.922853] evbug: Event. Dev: input0, Type: 0, Code: 0, Value: 0
[   47.482730] evbug: Event. Dev: input0, Type: 1, Code: 408, Value: 1
[   47.482838] evbug: Event. Dev: input0, Type: 0, Code: 0, Value: 0
[   48.332770] evbug: Event. Dev: input0, Type: 1, Code: 408, Value: 0
[   48.332879] evbug: Event. Dev: input0, Type: 0, Code: 0, Value: 0
[   48.922750] evbug: Event. Dev: input0, Type: 1, Code: 408, Value: 1
[   48.922856] evbug: Event. Dev: input0, Type: 0, Code: 0, Value: 0
[   49.712748] evbug: Event. Dev: input0, Type: 1, Code: 408, Value: 0
[   49.712854] evbug: Event. Dev: input0, Type: 0, Code: 0, Value: 0
```

**Unexpected behaviour you saw**
 > None

**Steps to reproduce the problem**
Define a gpio-keys based input in the DTS and use udevadm monitor to inspect for events when pressing the button.

**Comment by boucman** [comment]
udev is for hotplug events, not normal hardware events. 

Moreover, IIUC the gpio-keys driver hides the gpio event entirely and only produces an input event. so you need to detect it as any input event.

my guess is that udev is not the right tool for that. 

However, I _think_ the KEY_RESTART is handled by logind. (if you use it) so you might want to look in that direction

**Comment by robimarko** [comment]
Thanks for the tip.
I was under the impression that udev would work as I found multiple examples of GPIO based power buttons.

I looked through the docs and Googled around but I found no reference to KEY_RESTART and logind(I am using it).

Do you maybe have an example of how to handle an input event under systemd?
I am not used to it under embedded cases, so sorry for basic questions.

**Comment by boucman** [comment]
well... udev does the work of _classifying_ buttons... i.e it configures what key will be treated as a keyboard and what key will be a button. (that's what you found in the udev configuration)

But it doesn't deal with the event itself
I don't really understand how the input subsystem works, though... maybe look for libevdev and the various related tools ?

**Comment by robimarko** [comment]
Thanks, I will then have to look at using libevdev.
I try to avoid more dependecies

**Comment by poettering** [comment]
udev is not involved with handling key presses.

logind otoh watches for KEY_POWER, KEY_POWER2, KEY_SUSPEND and KEY_SLEEP keys, and triggers poweroff and system suspend in those case (it also watches SW_LID/SW_DOCK switches).

Hence: no code in systemd currently handles KEY_RESTART. We can add that of course. But I wasn't aware of the key event I must say. Is it generally accepted that KEY_RESTART means "reboot"?

**Comment by robimarko** [comment]
Thanks for the explanation, now it makes more sense.
Yeah, KEY_RESTART has been used for a quite while by OpenWrt and other projects to indicate reboot, as far as I am aware it has been introduced in the kernel for exactly the case of indicating reboot since only KEY_POWER was available at the time.

Do you maybe have a tip how would one would go to indicate factory reset with a button?

**Comment by poettering** [comment]
if KEY_RESET is accepted widely to mean reboot, we can certainly support that natively in logind. Will convert this bug into an RFE for that.

If you need a quick hack to make this work without a fix, no idea. Otherwise, of course we'd be delighted that adds KEY_RESET handling to logind, following how KEY_POWER is currently handled.

**Comment by robimarko** [comment]
Thanks,
I am now trying to find a way how to start a script of application if the button is held pressed for a couple of seconds.

**Comment by TriMoon** [comment]
@poettering you [mentioned](https://github.com/systemd/systemd/issues/16400#issuecomment-657763150) `KEY_RESET` while @robimarko was mentioning `KEY_RESTART`, i think they are different keys?

> Yeah, KEY_RESTART has been used for a quite while by OpenWrt and other projects to indicate reboot, as far as I am aware it has been introduced in the kernel for exactly the case of indicating reboot since only KEY_POWER was available at the time.
> 
> Do you maybe have a tip how would one would go to indicate factory reset with a button?

I think if there is a `KEY_RESET` key-code you could use that :wink:

**Comment by robimarko** [comment]
Nice catch, KEY_RESET does not exist in Linux, while KEY_RESTART does exist for a while.

**Comment by poettering** [comment]
So, if I look around the key code tables. It appears KEY_RESTART was introduced along with KEY_PLAY/KEY_STOP and things like that, i.e. was more about media playback.

If we'd blindly start making logind use it for rebooting this might end up being very weird if people actually have a lirc remote control or so with a restart key...

Wouldn't it be better to just submit a patch to the kernel to add a proper KEY_REBOOT and KEY_FACTORY_RESET or so to the list of known keys? If so we can support both safely in logind out of the box.

**Comment by robimarko** [comment]
Yes, it was introduced along but when you google around you will find that it has been widely used to indicate device restart.
It's used in 94 DTS in the upstream kernel for that purpose, also projects like Barebox have also added handlers for it to reboot the device.

https://elixir.bootlin.com/linux/latest/A/ident/KEY_RESTART

**Comment by poettering** [comment]
well, but does lirc generate it too for remote controls?

i mean, just because two different projects use it it doesn't mean there was no conflict...

**Comment by robimarko** [comment]
I am not aware of how Lirc works.
Even if there is a conflict, shouldn't the fact that Linux uses it exclusively to indicate device reboot and/or factory reset be more important than how downstream projects do things?

**Comment by poettering** [comment]
lirc is in the kernel, no (or some IR layer is, no)? i.e. i wonder about the "exclusively"...

**Comment by TriMoon** [comment]
I personally don't see a negative side to being able to reboot a machine using an IR-controller 😉

**Comment by poettering** [comment]
> I personally don't see a negative side to being able to reboot a machine using an IR-controller wink

Well, that would be fine, unless of course it's next to those play/pause buttons that are usually used for media playback.

So, looking at the kernel sources itself: `drivers/media/rc/keymaps/` has the IR keymaps the kernel knows about. It's quite a few, but only 3 of them use KEY_RESTART. OTOH there are many many more files in the linux source tree that apparently use it as a reboot key. Hence maybe we can accept the latter use won...

**Comment by robimarko** [comment]
I have tested KEY_RESTART handling in the same way as KEY_POWER and it works.
Is that acceptable so that I can start preparing the patch for submission?

Also, is there some logind logic that would allow launching custom userspace script of application on certain input event?

**Comment by poettering** [comment]
> I have tested KEY_RESTART handling in the same way as KEY_POWER and it works.
> Is that acceptable so that I can start preparing the patch for submission?

Sure, send a patch.

> Also, is there some logind logic that would allow launching custom userspace script of application on certain input event?

make the script a system wrapper and pull it in from reboot.target?

**Comment by keszybz** [comment]
Fixed in adbb2b6afed7858f64f5b940579f6efe79eee997.
